velox/exec/Spill.h

```diff
@@ -0,0 +1,340 @@
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include "velox/common/file/File.h"
+#include "velox/exec/TreeOfLosers.h"
+#include "velox/vector/ComplexVector.h"
+#include "velox/vector/DecodedVector.h"
+#include "velox/vector/FlatVector.h"
+
+namespace facebook::velox::exec {
+
+// A lightweight, non-owning reference to a row in a RowVector read from
+// spill.
+struct VectorRow {
+  std::vector<std::unique_ptr<DecodedVector>>* decoded;
+  RowVector* rowVector;
+  vector_size_t index;
+};
+
+// Input stream backed by spill file.
+class SpillInput : public ByteStream {
+ public:
+  // Reads from 'input' using 'buffer' for buffering reads.
+  SpillInput(std::unique_ptr<ReadFile>&& input, BufferPtr buffer)
+      : input_(std::move(input)), buffer_(std::move(buffer)) {
+    size_ = input_->size();
+    next(true);
+  }
+
+  void next(bool throwIfPastEnd) override;
+
+  // True if of the file has been read into vectors.
+  bool atEnd() const {
+    return offset_ == size_ && ranges()[0].position >= ranges()[0].size;
+  }
+
+  // Returns the offset of the first byte of input that is not yet consumed. May
+  // be used with seekp to rewind a read.
+  Position tellp() const override {
+    return std::make_tuple(
+        nullptr, offset_ - current_->size + current_->position);
+  }
+
+  // Positions the file so tht the byte at 'position' is the first to be read.
+  void seekp(Position position) override;
+
+ private:
+  std::unique_ptr<ReadFile> input_;
+  uint64_t size_;
+  BufferPtr buffer_;
+  uint64_t offset_ = 0;
+};
+
+// A source of VectorRows coming either from a file or memory.
+class SpillStream {
+ public:
+  SpillStream(TypePtr type, memory::MemoryPool& pool)
+      : type_(std::dynamic_pointer_cast<const RowType>(type)), pool_(pool) {
+    ordinal_ = ++ordinalCounter_;
+  }
+
+  virtual ~SpillStream() = default;
+
+  virtual bool atEnd() const = 0;
+
+  virtual std::string label() const {
+    return fmt::format("{}", ordinal_);
+  }
+
+  VectorRow next() {
+    if (index_ == numRows_) {
+      nextBatch();
+      index_ = 0;
+      numRows_ = rowVector_->size();
+      SelectivityVector allRows(numRows_);
+      auto width = rowVector_->childrenSize();
+      if (decoded_.size() < width) {
+        decoded_.resize(width);
+      }
+      for (auto i = 0; i < width; ++i) {
+        if (!decoded_[i]) {
+          decoded_[i] = std::make_unique<DecodedVector>();
+        }
+        decoded_[i]->decode(*rowVector_->childAt(i), allRows);
+      }
+    }
+    return {&decoded_, rowVector_.get(), index_++};
+  }
+
+  virtual uint64_t size() const {
```

Comment:
This method always returns zero. Is this intentional? Should it be defined as = 0 to force derived classes to provide an implementation?

Reply:
I still have this question:

----
This method always returns zero. Is this intentional? Should it be defined as = 0 to force derived classes to provide an implementation?